### PR TITLE
feat(server): add transportMode

### PR DESCRIPTION
**Pull request by Loonride**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Yes, based on moving around and changing existing `clientMode` and `serverMode` tests

### Motivation / Use-Case

https://github.com/webpack/webpack-dev-server/issues/1860

This replaces `clientMode` and `serverMode` options with the usage:

```
transportMode: {
  server: 'ws',
  client: 'ws'
}
```

### Breaking Changes

Experimental options `clientMode` and `serverMode` removed

### Additional Info

- `transportMode-option.test.js` is bulky and maybe should be broken down
- the configuration I added into `Server.js` could definitely be moved into some unified `updateOptions` helper to be tested more easily.

### CODE REVIEW

File: lib/Server.js [hunk 1]
```diff
@@ -66,20 +66,29 @@ class Server {
...
+        default:
+          this.options.transportMode.server =
+            this.options.transportMode.server || 'sockjs';
+          this.options.transportMode.client =
+            this.options.transportMode.client || 'sockjs';
+      }
+
       this.log.warn(
-        'clientMode is an experimental option, meaning its usage could potentially change without warning'
+        'transportMode is an experimental option, meaning its usage could potentially change without warning'
```
Comment (evilebottnawi) [inline lib/Server.js]: Maybe we can move this in util, like `getTransportMode(options: object): object`

**Comment by Loonride** [comment]
@evilebottnawi @hiroppy What is the cause of the Lint `npm audit` problem?

**Comment by evilebottnawi** [comment]
@Loonride i think we need update `package-json.lock`

**Comment by hiroppy** [comment]
I created a pr which updated package-lock.json. https://github.com/webpack/webpack-dev-server/pull/2123

**Comment by jamesgeorge007** [comment]
Guys can you let me know how you're planning to handle #2120

**Comment by jamesgeorge007** [comment]
@evilebottnawi @Loonride @hiroppy IMO merging #2120 before this one doesn't impose any harm right :thinking:

**Comment by jamesgeorge007** [comment]
@evilebottnawi @Loonride @hiroppy IMO merging #2120 before this one doesn't impose any harm right :thinking:

**Comment by Loonride** [comment]
@evilebottnawi Looks like unlucky e2e test run. Maybe there is a way to let `jest` run concurrently for all other tests, but then run e2e one at a time? Now that I look at it, I see `Client.test.js` and `Progress.test.js` are nearby each other in the test run failures, and they are both editing the same file, which should be changed. I will send another PR to improve it.

**Comment by Loonride** [comment]
@evilebottnawi Ready for merge.

**Comment by evilebottnawi** [comment]
@Loonride we need something any for CI fix?

**Comment by Loonride** [comment]
> @Loonride we need something any for CI fix?

The CI problems are not due to the changes here. I was trying to fix these e2e issues in this PR: https://github.com/webpack/webpack-dev-server/pull/2163, but still not perfect unfortunately.

**Comment by evilebottnawi** [comment]
@Loonride we can merge without CI green, maybe i can help?

**Comment by Loonride** [comment]
> we can merge without CI green

Yes, I think this PR is good and we can do CI improvements in a separate PR.

> maybe i can help?

With CI e2e improvements? Yes. Maybe we should make an issue to work through it.

**Comment by evilebottnawi** [comment]
@Loonride 
> With CI e2e improvements? Yes. Maybe we should make an issue to work through it.

I will look on this tomorrow :+1:

**Pull request merged**